### DISCUSSION TITLE
Feature Request: using IFC AdvancedSweptSolid to sweep user-defined 2D profile along arbitrary 3D polyline

### DISCUSSION THREAD

**Discussion by swiss-knight** [IfcOpenShell/IfcOpenShell#5724] [Ideas]
Hello 👋,

As an {architect | civil engineer} given:

> 
AdvancedSweptSolid | swept area solids created by sweeping a profile along a directrix, and tapered sweeps
-- | --

in the [current IFC4X3_ADD2 spec](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcShapeRepresentation.htm), I would love to be able to use the [ifcopenshell.util.shape_builder](https://docs.ifcopenshell.org/autoapi/ifcopenshell/util/shape_builder/index.html#module-ifcopenshell.util.shape_builder) in order to sweep a 2D shape along a 3D curve. 

Exactly as one can currently use the [IfcSweptDiskSolid](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcSweptDiskSolid.htm) implemented with the current [`create_swept_disk_solid`](https://docs.ifcopenshell.org/autoapi/ifcopenshell/util/shape_builder/index.html#ifcopenshell.util.shape_builder.ShapeBuilder.create_swept_disk_solid) method.

This would be extremely useful for creating complex 3D structures.

E.g.:

```py
import ifcopenshell
import ifcopenshell.api.context
import ifcopenshell.api.geometry
import ifcopenshell.api.project
import ifcopenshell.api.root
import ifcopenshell.api.unit
from ifcopenshell.util.shape_builder import V

model = ifcopenshell.api.project.create_file(version="IFC4X3")
ifcopenshell.api.root.create_entity(model, ifc_class="IfcProject", name='foo')
model3d = ifcopenshell.api.context.add_context(model, context_type="Model")
body = ifcopenshell.api.context.add_context(
    model,
    context_type="Model",
    context_identifier="Body",
    target_view="MODEL_VIEW",
    parent=model3d,
)
builder = ifcopenshell.util.shape_builder.ShapeBuilder(ifc_file=model)
path = [(1,2,3),(2,2,4),(3,4,6),...] # list of 3D cartesian coordinates
curve = builder.polyline(path)

# Using a user-defined profile
profile = model.create_entity("IfcRectangleProfileDef", ProfileName="600x300", ProfileType="AREA", XDim=600, YDim=300)

# Using for example a new create_swept_profile_solid() method to build a 3D solid:
swept_curve = builder.create_swept_profile_solid(curve, profile=profile)
```
(Maybe this could be inspired by: [IfcSweptAreaSolid](https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcSweptAreaSolid.htm))

Thanks a lot amazing IfcOpenShell folks! 👍

**Comment by Xema-Pastor** [comment]
Hi all

I join the request, but adding a tapering feature in order to create tapered solids. Please, develope the IfcExtrudedAreaSolidTapered and IfcRevolvedAreaSolidTapered classes.

Thank you
